src/gil.rs

```diff
@@ -241,37 +239,68 @@ impl GILPool {
 impl Drop for GILPool {
     fn drop(&mut self) {
         unsafe {
-            let pool = POOL.get_or_init();
-            pool.drain(self.python(), self.owned);
+            OWNED_OBJECTS.with(|owned_objects| {
+                // Note: inside this closure we must be careful to not hold a borrow too long, because
+                // while calling Py_DECREF we may cause other callbacks to run which will need to
+                // register objects into the GILPool.
+                let len = owned_objects.borrow().len();
```

Comment:
Too inefficient?

Reply:
Hmm :thinking: 
If we need this kind of hack, we should consider simply lock other threads.

Reply:
The problem isn't other threads, but that multiple `GILPool`s can exist in the same thread. They might even be created & destroyed while this `GILPool`'s drop is running.

For example, when `Py_DECREF` causes a reference count to go to zero, then we may end up running a `tp_dealloc` callback, which might create a new `GILPool`, which might run other python code, which might run something else which creates another `GILPool`, ... (and so on). I ***think*** that all those `GILPool`s should be cleaned up by the time the `tp_dealloc` callback ends, but I was feeling cautious.

If we are concerned about performance we should write some more benchmarks to compare https://github.com/PyO3/pyo3/pull/887 against pyo3 master? If there's a noticeable performance regression then we could consider using `UnsafeCell` instead of `RefCell`. I'd prefer avoid the `unsafe` unless we really need it though.

Reply:
Another alternative is maybe for each `GILPool` to own its own `Vec` storage. We could even try using [`smallvec`](https://docs.rs/smallvec/1.4.0/smallvec/) which might mean that most `GILPool`s will never need an allocation?

Reply:
EDIT: thinking about this more, we already use `UnsafeCell` in this file. So, maybe we should just proceed and use `UnsafeCell` here too.

Reply:
(Do you want me to open another PR to change `RefCell` -> `UnsafeCell`?)

Reply:
Haha, one last thought, sorry :D

Actually I'm not sure that `UnsafeCell` is sound. Any `register_owned` owned call might cause the `Vec` to resize, which could move the contents of the `Vec` storage. So we have to be ***very*** careful to make sure we're using the shared `Vec` correctly, else we're accessing invalid memory.

This problem also exists without `thread_local`, I guess. Just now that we're looking into this file we see the dangers of the existing implementation!

Reply:
> If we are concerned about performance

I don't mind performance so much here, but now I don't feel this solution so *reliable*.

Reply:
My temporal solution: https://github.com/PyO3/pyo3/pull/887/commits/75c807f31d8dd478b00420abee9183851b4b7239